libs/core/tests/unit_tests/callbacks/test_async_callback_manager.py

```diff
@@ -0,0 +1,84 @@
+import asyncio
+import pytest
+import contextvars
+from uuid import UUID
+from typing import Any, Optional
+from langchain_core.callbacks import AsyncCallbackManager, AsyncCallbackHandler
+from contextlib import asynccontextmanager
+
+counter_var = contextvars.ContextVar("counter", default=0)
+
+shared_stack = []
+
+
+@asynccontextmanager
+async def set_counter_var():
+    token = counter_var.set(0)
+    try:
+        yield
+    finally:
+        counter_var.reset(token)
+
+
+class StatefulAsyncCallbackHandler(AsyncCallbackHandler):
+    def __init__(self, name: str, delay: float = 0.1, run_inline: bool = True):
+        self.name = name
+        self.delay = delay
+        self.run_inline = run_inline
+
+    async def on_llm_start(
+        self,
+        serialized: dict[str, Any],
+        prompts: list[str],
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        **kwargs: Any,
+    ) -> None:
+        event_name = "on_llm_start"
+        prompt = prompts[0]
+        await asyncio.sleep(self.delay)
```

Comment:
Is most of the test case necessary?  Can we capture what we need in ~10 lines of code verifying that context vars is being set and can be read?

Reply:
@eyurtsev , Thank you for your questions. After careful consideration:

1. The `delay` is unnecessary and can be removed without affecting the test's ability to demonstrate the issue. I've done this now.

2. Regarding simplifying the test case, it's important to note that a simpler test using only `contextvars` would actually pass, even if the `run_inline` attribute is not being honored. This is because `contextvars` maintain context across async tasks but don't reflect execution order or concurrency issues.

The test case using `shared_stack` is necessary because:

- It reveals concurrent execution issues that `contextvars` alone cannot detect.
- It demonstrates the need to honor `run_inline=True` in `AsyncCallbackManager`.
- It highlights potential problems in scenarios where handlers rely on sequential execution.

While we could simplify some aspects of the test, maintaining the core structure with `shared_stack` is crucial for effectively demonstrating the issue. I highlighted this in the PR description, but now I'm unsure how to simplify it further while still demonstrating the issue. I'm open to suggestions on how to make it more concise while still effectively capturing the problem.